File: wasm/src/pathfinding/breadth_first_search.rs
pub fn breadth_first_search(
  blocked_cells_numbers: &[u8],
  w: usize,
  h: usize,
  start: usize,
  end: usize,
  diagonal: bool,
) -> Vec<usize> {
  let mut queue: Vec<usize> = vec![start];
  let mut queue_start = 0; // pointer for the start of the queue

  let mut parents: Vec<isize> = vec![-2; w * h]; // Using a vector instead of a map
  parents[start] = -1; // Start has no parent

  while queue_start < queue.len() {
    let current_cell = queue[queue_start];
    queue_start += 1;

    if current_cell == end {
      return reconstruct_path(&parents, end, start);
    }

    let px = current_cell % w;
    let py = current_cell / w;

    if px > 0 && blocked_cells_numbers[current_cell - 1] != 1 && parents[current_cell - 1] == -2 {
      // left
      queue.push(current_cell - 1);
      parents[current_cell - 1] = current_cell as isize;
    }
    if px < w - 1 && blocked_cells_numbers[current_cell + 1] != 1 && parents[current_cell + 1] == -2 {
      // right
      queue.push(current_cell + 1);
      parents[current_cell + 1] = current_cell as isize;
    }
    if py > 0 && blocked_cells_numbers[current_cell - w] != 1 && parents[current_cell - w] == -2 {
      // up
      queue.push(current_cell - w);
      parents[current_cell - w] = current_cell as isize;
    }
    if py < h - 1 && blocked_cells_numbers[current_cell + w] != 1 && parents[current_cell + w] == -2 {
      // down
      queue.push(current_cell + w);
      parents[current_cell + w] = current_cell as isize;
    }

    if diagonal {
      if px > 0 && py > 0 && parents[current_cell - w - 1] == -2 &&
        blocked_cells_numbers[current_cell - w - 1] != 1 &&
        blocked_cells_numbers[current_cell - 1] != 1 &&
        blocked_cells_numbers[current_cell - w] != 1
      {
        // top-left
        queue.push(current_cell - w - 1);
        parents[current_cell - w - 1] = current_cell as isize;
      }
      if px < w - 1 && py > 0 && parents[current_cell - w + 1] == -2 &&
        blocked_cells_numbers[current_cell - w + 1] != 1 &&
        blocked_cells_numbers[current_cell + 1] != 1 &&
        blocked_cells_numbers[current_cell - w] != 1
      {
        // top-right
        queue.push(current_cell - w + 1);
        parents[current_cell - w + 1] = current_cell as isize;
      }
      if px > 0 && py < h - 1 && parents[current_cell + w - 1] == -2 &&
        blocked_cells_numbers[current_cell + w - 1] != 1 &&
        blocked_cells_numbers[current_cell - 1] != 1 &&
        blocked_cells_numbers[current_cell + w] != 1
      {
        // bottom-left
        queue.push(current_cell + w - 1);
        parents[current_cell + w - 1] = current_cell as isize;
      }
      if px < w - 1 && py < h - 1 && parents[current_cell + w + 1] == -2 &&
        blocked_cells_numbers[current_cell + w + 1] != 1 &&
        blocked_cells_numbers[current_cell + 1] != 1 &&
        blocked_cells_numbers[current_cell + w] != 1
      {
        // bottom-right
        queue.push(current_cell + w + 1);
        parents[current_cell + w + 1] = current_cell as isize;
      }
    }
  }

  vec![]
}

// Reconstruct the path using a parent array
fn reconstruct_path(parents: &[isize], end: usize, start: usize) -> Vec<usize> {
  let mut path: Vec<usize> = Vec::new();
  let mut current = end as isize;

  while current != -1 {
    path.push(current as usize);
    if current == start as isize {
      break;
    }
    current = parents[current as usize];
  }

  path.reverse();
  path
}

File: website/script/pathfinding/breadth-first-search.ts
export function breadthFirstSearch(
  blockedCellsNumbers: Uint8Array,
  w: number,
  h: number,
  start: number,
  end: number,
  diagonal: boolean = false
): number[] {
  const gridSize: number = w * h;
  const bufferSize: number = (gridSize << 3) + gridSize;
  const buffer: ArrayBuffer = new ArrayBuffer(bufferSize);
  const queue: Uint32Array = new Uint32Array(buffer, 0, gridSize);
  const parents: Uint32Array = new Uint32Array(buffer, gridSize * 4, gridSize);
  parents.fill(0xFFFF);
  const blockedCells: Uint8Array = new Uint8Array(buffer, gridSize * 8, gridSize);
  blockedCells.set(blockedCellsNumbers);

  queue[0] = start;
  parents[start] = -1;

  const maxX: number = w - 1;
  const maxY: number = h - 1;
  
  let queueLength: number = 1;
  let queueStart: number = 0;

  while (queueStart < queueLength) {
    // @ts-expect-error
    const currentCell: number = queue[queueStart++];
    if (currentCell === end) {
      const path: number[] = [];
      let current: number = end;
      while (current && current !== -1) {
        path.push(current);
        if (current === start) break;
        // @ts-expect-error
        current = parents[current];
      }
      return path.reverse();
    }
    const px: number = currentCell % w;
    const py: number = ~~(currentCell / w);

    const left: number = currentCell - 1;
    const right: number = currentCell + 1;
    const top: number = currentCell - w;
    const bottom: number = currentCell + w;

    const canMoveLeft: boolean = px > 0;
    const canMoveRight: boolean = px < maxX;
    const canMoveUp: boolean = py > 0;
    const canMoveDown: boolean = py < maxY;

    if (
      canMoveLeft &&
      parents[left] === 0xFFFF &&
      blockedCells[left] !== 1
    ) { // left
      queue[queueLength++] = left;
      parents[left] = currentCell;
    }
    if (
      canMoveRight &&
      parents[right] === 0xFFFF &&
      blockedCells[right] !== 1
    ) { // right
      queue[queueLength++] = right;
      parents[right] = currentCell;
    }
    if (
      canMoveUp &&
      parents[top] === 0xFFFF &&
      blockedCells[top] !== 1
    ) { // up
      queue[queueLength++] = top;
      parents[top] = currentCell;
    }
    if (
      canMoveDown &&
      parents[bottom] === 0xFFFF &&
      blockedCells[bottom] !== 1
    ) { // down
      queue[queueLength++] = bottom;
      parents[bottom] = currentCell;
    }

    if (diagonal) {
      let topLeft: number = currentCell - maxX;
      let topRight: number = currentCell - w + 1;
      let bottomLeft: number = currentCell + maxX;
      let bottomRight: number = currentCell + w + 1;
      if (
        // @ts-expect-error
        canMoveUp & canMoveLeft &&
        parents[topLeft] === 0xFFFF &&
        blockedCells[topLeft] !== 1 &&
        blockedCells[top] !== 1 &&
        blockedCells[left] !== 1
      ) { // top-left
        queue[queueLength++] = topLeft;
        parents[topLeft] = currentCell;
      }
      if (
        // @ts-expect-error
        canMoveUp & canMoveRight &&
        parents[topRight] === 0xFFFF &&
        blockedCells[topRight] !== 1 &&
        blockedCells[top] !== 1 &&
        blockedCells[right] !== 1
      ) { // top-right
        queue[queueLength++] = topRight;
        parents[topRight] = currentCell;
      }
      if (
        // @ts-expect-error
        canMoveDown & canMoveLeft &&
        parents[bottomLeft] === 0xFFFF &&
        blockedCells[bottomLeft] !== 1 &&
        blockedCells[bottom] !== 1 &&
        blockedCells[left] !== 1
      ) { // bottom-left
        queue[queueLength++] = bottomLeft;
        parents[bottomLeft] = currentCell;
      }
      if (
        // @ts-expect-error
        canMoveDown & canMoveRight &&
        parents[bottomRight] === 0xFFFF &&
        blockedCells[bottomRight] !== 1 &&
        blockedCells[bottom] !== 1 &&
        blockedCells[right] !== 1
      ) { // bottom-right
        queue[queueLength++] = bottomRight;
        parents[bottomRight] = currentCell;
      }
    }
  }

  return [];
}
